Create Topic Editor (Part 2): Created Topics and Skills dashboard

aks681 opened the pull request:
In this PR, a basic topics and skills dashboard is created. This PR builds on PR #5090 .
The commits after (and including) 7034a6c, are the additions done for this PR.
Also, to view the actual dashboard, ENABLE_NEW_STRUCTURES must be changed to True in feconf.py.

aks681 committed:
Merged with frontend-domain-objects

aks681 committed:
added topics and skills dashboard

aks681 committed:
fixed errors

aks681 committed:
minor change

seanlip commented on core/templates/dev/head/pages/topics_and_skills_dashboard/topics_and_skills_dashboard.html:
```diff
@@ -1,2 +1,227 @@
+    </div>
+  </div>
+  <style>
+  .oppia-dashboard-tabs {
```
Indent all these by 2, since it's in an inner scope of the "style" tag.

(Was this a copy-paste issue? If so, would you mind fixing the original source of this pattern too?)

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/pages/topics_and_skills_dashboard/topics_and_skills_dashboard.html:
```diff
@@ -1,2 +1,227 @@
+    padding-top: 7%;
+    border-bottom: 1px solid #eee;
+    display: -webkit-flex;
+    display: flex;
```
Within each CSS rule, alphabetize attrs (so, e.g., border-bottom then display then flex-wrap, etc.)

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/pages/topics_and_skills_dashboard/topics_and_skills_dashboard.html:
```diff
@@ -1,2 +1,227 @@
+  }
+
+  a:hover {
+    text-decoration: none !important;
```
Do not use !Important, and do not use a rule that extends beyond the scope of this page (e.g. to "a" tags everywhere). Every CSS rule on this page should have a prefix that scopes it to just this page. Shared CSS should go in oppia.css.

@apb7 -- any guidance to add here? (Lint checks should be catching stuff like this.) Also /cc @kevinlee12

seanlip replied:
(Also @apb7 @kevinlee12 please let's file issues for these CSS lint checks if we're not tracking them already.)

apb7 replied:
Hi @seanlip! We do have [`declaration-no-important`](https://stylelint.io/user-guide/rules/declaration-no-important/) rule but it has not been enabled since it was not a part of the recommended rules.
Regarding "Every CSS rule on this page should have a prefix that scopes it to just this page.", we require a custom rule.
I am filing an issue to extend the CSS rules and listing these two rules.
Thanks!

seanlip replied:
OK, thanks! Let's file one also for alphabetization (see https://github.com/oppia/oppia/pull/5094/files/3ec3a53166e10481f5215eaf08939f20a6cc7c2f..cdf5d3df4aa5d3f7f1e99953eaacc966a145985a#r194923524).

Also, we need one for indentation in HTML? https://github.com/oppia/oppia/pull/5094/files/3ec3a53166e10481f5215eaf08939f20a6cc7c2f..cdf5d3df4aa5d3f7f1e99953eaacc966a145985a#r194923479

apb7 replied:
Shall I add them to the same issue as they are related to CSS rules only?

aks681 replied:
Done, removed it.

seanlip replied:
@apb7 sure, go ahead. (As long as they're kept track of, I don't really mind where; that's up to you.)

seanlip commented on core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js:
```diff
@@ -75,6 +75,9 @@ oppia.directive('topNavigationBar', [
           $scope.userMenuIsShown = ($scope.NAV_MODE !== NAV_MODE_SIGNUP);
           $scope.standardNavIsShown = (
             NAV_MODES_WITH_CUSTOM_LOCAL_NAV.indexOf($scope.NAV_MODE) === -1);
+          if ($scope.NAV_MODE === 'topics_and_skills_dashboard') {
```
Why not just include this in NAV_MODES_WITH_CUSTOM_LOCAL_NAV?

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/pages/topics_and_skills_dashboard/TopicsAndSkillsDashboard.js:
```diff
@@ -0,0 +1,43 @@
+      function(response) {
+        $scope.topicSummaries = response.data.topic_summaries;
+        $scope.skillSummaries = response.data.skill_summaries;
+        $scope.activeTab = 'topics';
```
Let's make topics a constant -- perhaps TAB_NAME_TOPICS.

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/pages/topics_and_skills_dashboard/TopicsAndSkillsDashboard.js:
```diff
@@ -0,0 +1,43 @@
+      function(errorResponse) {
+        if (FATAL_ERROR_CODES.indexOf(errorResponse.status) !== -1) {
+          AlertsService.addWarning('Failed to get dashboard data');
+        }
```
else error somehow?

aks681 replied:
FATAL_ERROR_CODES has 400, 401, 404 and 500. So, would an else be needed? or shall I remove the if condition entirely?

seanlip replied:
Could remove the if condition, or do something like `else "unexpected error code from server"`. Probably want to be noisy somehow if that sort of error shouldn't happen in the first place, I guess.

aks681 replied:
Done, added an else statement.

seanlip commented on core/templates/dev/head/pages/topics_and_skills_dashboard/topics_and_skills_dashboard.html:
```diff
@@ -1,2 +1,227 @@
+    <div ng-if="topicSummaries.length > 0 || skillSummaries.length > 0">
+      <ul class="oppia-dashboard-tabs">
+        <li ng-if="topicSummaries.length > 0"
+            ng-class="{'oppia-dashboard-tabs-active': activeTab === 'topics'}">
```
Here and elsewhere below, see if you can use constants for the tab names.

aks681 replied:
Done.

seanlip reviewed (commented):
@aks681 I took a full pass at this, it looks generally good. Left some inline comments. Thanks!

aks681 committed:
fixed merge conflict and made review changes

aks681 committed:
changed card view to list view

aks681 commented:
@seanlip PTAL. Added a total_skill_count field to topic_summary_model and worked_examples_count field to skill_Summaries_model. Also, changed the dashboard structure from cards to a tabular view:
![screenshot from 2018-06-13 23-57-54](https://user-images.githubusercontent.com/22347970/41370675-e261f7aa-6f65-11e8-8c16-e8b6d391f753.png)

aks681 committed:
minor changes

aks681 committed:
made review change

aks681 committed:
disabled new structures constant

seanlip commented on assets/i18n/en.json:
```diff
@@ -501,14 +509,16 @@
     "I18N_TEACH_PAGE_PUBLICATION_POLICY_PARAGRAPH_2": "Explorations are meant to be continuously improved, and explorations that haven't been edited for a long time may be considered \"orphaned\". In this case, ownership may be handed to the Oppia community at large (at moderator discretion), so that the exploration can continue to be improved.",
     "I18N_TEACH_PAGE_TAB_PARTICIPATION_PLAYBOOK": "Participation Playbook",
     "I18N_TEACH_PAGE_TAB_TEACH_WITH_OPPIA": "Teach with Oppia",
+    "I18N_TOPICS_AND_SKILLS_DASHBOARD_INTRO_MESSAGE_PART_ONE": "It looks like no topics or skills have been created yet.",
```
I think "No topics or skills have been created yet." suffices (and you don't need line 2). There's probably no need to encourage the admin to create something, and it's one less line for translators to handle.

Also, generally, we only translate learner-facing pages. Is this page learner-facing?

aks681 replied:
No, it's not learner facing.

aks681 replied:
Done, changed the text.

seanlip commented on core/templates/dev/head/pages/topics_and_skills_dashboard/topics_and_skills_dashboard.html:
```diff
@@ -23,195 +23,286 @@ <h2 class="oppia-dashboard-title" translate="I18N_DASHBOARD_TOPICS_AND_SKILLS_DA
-
-    .oppia-dashboard-container .oppia-dashboard-tabs li {
-      width: auto;
+    .oppia-dashboard-list-view-item {
```
One issue with calling this ".oppia-dashboard-..." is that there are at least three dashboards in Oppia at the moment (creator, learner, topics-and-skills). So, if this is meant for all these three dashboards, then maybe we should put it in oppia.css. Otherwise we need a clearer way to scope it to this page only.

It's true though that callling this .oppia-topics-and-skills-dashboard-list-view-item is a bit of a mouthful :) Here's an alternative: perhaps add an oppia-topics-and-skills-dashboard class to the topmost tag of the content? Then you can call the CSS stuff like ".oppia-topics-and-skills-dashboard .list-view-item".

Even better, try to have this controller HTML be minimal. Use directives instead for the main components, and let those directives have their own HTML/CSS. Prefix each CSS tag in the directive with the name of the directive. That is probably the best way and it keeps things modular.

aks681 replied:
Done, split into 2 directives for topics list and skills list.

seanlip reviewed (commented):
Thanks @aks681! The table looks great :) Just a couple comments.

aks681 committed:
made review changes

aks681 committed:
fixed  errors

aks681 committed:
disabled new structures

aks681 commented:
@seanlip PTAL. Created the directives. Also, made the ' ( 1 uncategorized ) ' in the above mock disappear when in small windows, since that was overlapping with the row below.

seanlip commented on assets/i18n/en.json:
```diff
@@ -93,17 +93,28 @@
     "I18N_DASHBOARD_EXPLORATIONS_SORT_BY_TOTAL_PLAYS": "Total Plays",
     "I18N_DASHBOARD_EXPLORATIONS_SORT_BY_UNRESOLVED_ANSWERS": "Unresolved Answers",
     "I18N_DASHBOARD_OPEN_FEEDBACK": "Open Feedback",
+    "I18N_DASHBOARD_SKILLS": "Skills",
```
Ah -- there may have been a misunderstanding. I think we should remove strings we aren't planning to translate, so that we do not unnecessarily burden the translatewiki folks. Could we do that?

aks681 replied:
Oh, ok, except the title in the navbar, removed everything else from translations.

seanlip commented on core/templates/dev/head/pages/topics_and_skills_dashboard/SkillsListDirective.js:
```diff
@@ -0,0 +1,38 @@
+// limitations under the License.
+
+/**
+ * @fileoverview Controller for the navbar breadcrumb of the collection editor.
```
Ahem.

aks681 replied:
Oh, missed that, changed.

seanlip commented on core/templates/dev/head/pages/topics_and_skills_dashboard/SkillsListDirective.js:
```diff
@@ -0,0 +1,38 @@
+    return {
+      restrict: 'E',
+      scope: {
+        skillSummaries: '&'
```
Better: `getSkillSummaries: '&skillSummaries'`

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/pages/topics_and_skills_dashboard/TopicsListDirective.js:
```diff
@@ -0,0 +1,41 @@
+    return {
+      restrict: 'E',
+      scope: {
+        topicSummaries: '&'
```
Ditto.

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/pages/topics_and_skills_dashboard/skills_list_directive.html:
```diff
@@ -0,0 +1,84 @@
+<div class="skills-list">
+  <md-card layout="row" class="list-view-item">
+    <table class="dashboard-table">
+        <colgroup>
```
This and the next four lines: deindent by 2

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/pages/topics_and_skills_dashboard/skills_list_directive.html:
```diff
@@ -0,0 +1,84 @@
+</div>
+
+<style>
+  .skills-list .list-view-item {
```
Not sure you need the parent class. Wouldn't `skills-list .list-view-item` work? This is because we have a directive now so everything is in a `<skills-list>` tag?

aks681 replied:
Yeah, it does. Done.

seanlip commented on core/templates/dev/head/pages/topics_and_skills_dashboard/skills_list_directive.html:
```diff
@@ -0,0 +1,84 @@
+  .skills-list .list-view-item {
+    background: #fff;
+    margin: 20px 7.5px 0px 7.5px;
+    margin-left: 12.5%;
```
Why set both a margin and a margin-left?

aks681 replied:
Done, added the margin-left value to margin.

seanlip commented on core/templates/dev/head/pages/topics_and_skills_dashboard/topics_and_skills_dashboard.html:
```diff
@@ -16,123 +16,42 @@
-          </tr>
-        </table>
-      </md-card>
+    <div ng-if="activeTab === TAB_NAME_TOPICS">
```
Something's up with the indentation here, it does't look like a sibling of the previous line (though it is).

/cc @apb7 @kevinlee12 we should catch this automatically.

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/pages/topics_and_skills_dashboard/topics_and_skills_dashboard.html:
```diff
@@ -312,6 +199,8 @@ <h2 class="oppia-dashboard-title" translate="I18N_DASHBOARD_TOPICS_AND_SKILLS_DA
+  <script src="{{TEMPLATE_DIR_PREFIX}}/pages/topics_and_skills_dashboard/TopicsListDirective.js"></script>
+  <script src="{{TEMPLATE_DIR_PREFIX}}/pages/topics_and_skills_dashboard/SkillsListDirective.js"></script>
 
   <script   src="{{TEMPLATE_DIR_PREFIX}}/domain/topics_and_skills_dashboard/TopicsAndSkillsDashboardBackendApiService.js"></script>
```
Extra spaces before src.

For tidiness, maybe arrange the imports from pages/topics_and_skills_dashboard/ in alphabetical order by filename?

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/pages/topics_and_skills_dashboard/topics_list_directive.html:
```diff
@@ -0,0 +1,98 @@
+<div class="topics-list">
+  <md-card layout="row" class="list-view-item">
+    <table class="dashboard-table">
+        <colgroup>
```
ditto, indentation

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/pages/topics_and_skills_dashboard/topics_list_directive.html:
```diff
@@ -0,0 +1,98 @@
+</div>
+
+<style>
+  .topics-list .list-view-item {
```
ditto, not sure you even need a top-level class here; use the name of the directive

aks681 replied:
Done.

seanlip reviewed (changes requested):
The directives look great! Left some more comments.

aks681 committed:
made review changes

aks681 committed:
fixed errors

aks681 commented on core/templates/dev/head/domain/topic/TopicObjectFactory.js:
```diff
@@ -290,6 +290,18 @@ oppia.factory('TopicObjectFactory', ['SubtopicObjectFactory',
+
+    // Create a new, empty topic. This is not guaranteed to pass validation
+    // tests.
+    Topic.createEmptyTopic = function() {
```
When I was doing the topiceditor, I saw that in collections, when the StateService is initialised, an empty object was initialised, after which when the collection is set, by calling _setCollection() the collection is actually loaded, so this would be needed.

seanlip replied:
Ah OK. In that case, leave a comment about this here please. Perhaps:

```
// Create a new, empty topic that is not guaranteed to pass validation tests. This should only be used
// when initializing the topic editor (prior to setting it with data from the backend).
```

That said ... this seems a little weird to me. Why do we have this intermediate phase where the object is invalid? It doesn't seem to me that a blank topic is much better than a null value. I'll send an email to @BenHenning to check, but it's fine to keep this in parity with the collection editor for now.

aks681 replied:
Ok.

BenHenning replied:
Followed up via email, but to summarize on here also for posterity:

It's been a while since that code was changed, but I think the idea is that a null collection needs to be special cased, but an empty collection can still be valid. It's not clear to me if that invariant actually holds here, but given that we don't check anywhere if collection == empty collection (compared to checking if it's null), I suspect it does. This seems safer in general since it means that the editor needs to be properly support empty collections, and that in turn offers more predictability: the editor goes from showing a default empty collection to a properly loaded collection.

If the empty collection doesn't pass validation tests, that may be an oversight. Empty objects should be valid insomuch as they can be displayed properly, but it may not be valid to save them (that's okay--starting off with an empty collection is valid even if it can't be saved since it provides an initial space to start working).

To put another way, whatever default value is used in editors should satisfy:
- Facilitating some sort of default view without requiring special casing and without errors
- Not being saveable without creator interaction (this holds true for explorations and possibly collections--not as sure on the latter)

We could probably do a better job here, but I think that's the rough rational behind not using null.

aks681 committed:
disabled new structures

seanlip commented on core/templates/dev/head/domain/topic/TopicObjectFactory.js:
```diff
@@ -79,18 +79,34 @@ oppia.factory('TopicObjectFactory', ['SubtopicObjectFactory',
-    // in general, use topic.addSubtopic()
+    /**
+     * @param {number} id - The id of the subtopic that was deleted.
+     * @param {string} title - The title of the subtopic that w as deleted.
```
typo: w as --> was

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/domain/topic/TopicUpdateService.js:
```diff
@@ -97,30 +97,13 @@ oppia.factory('TopicUpdateService', [
-      });
-      return newId[0];
+    var _checkIfSkillIdInList = function(skillId, skillList) {
+      for (var i = 0; i < skillList.length; i++) {
```
Actually ... wouldn't `skillList.indexOf(skillId) !== -1` suffice?

also, I think skillList is misnamed. If its elements are skill IDs then it should be called skillIdList. In general, be very careful to distinguish between skill IDs and skills (or IDs and objects in general); they are different things. Make sure to double-check the variable names used in the callers, too.

aks681 replied:
Yes, ok, done.

seanlip commented on core/templates/dev/head/pages/topics_and_skills_dashboard/skills_list_directive.html:
```diff
@@ -1,82 +1,79 @@
 
 <style>
-  .skills-list .list-view-item {
+  skills-list .list-view-item {
```
Just to check, why do you need the .list-view-item class at all -- wouldn't just "skills-list" suffice"?

aks681 replied:
I tried by removing list-view-item, but then the styling was removed.

seanlip commented on core/templates/dev/head/pages/topics_and_skills_dashboard/skills_list_directive.html:
```diff
@@ -1,82 +1,79 @@
     background: #fff;
-    margin: 20px 7.5px 0px 7.5px;
-    margin-left: 12.5%;
+    margin: 20px 7.5px 0px 12.5%;
```
Prefer 0 to 0px. /cc @apb7 @kevinlee12

aks681 replied:
Done.

apb7 replied:
Added to Issue #5097. Thanks!

seanlip commented on core/templates/dev/head/pages/topics_and_skills_dashboard/topics_and_skills_dashboard.html:
```diff
@@ -25,38 +25,35 @@ <h2 class="dashboard-title" translate="I18N_DASHBOARD_TOPICS_AND_SKILLS_DASHBOAR
           <a class="dashboard-tabs-text"
-             ng-click="setActiveTab(TAB_NAME_TOPICS)"
-             translate="I18N_DASHBOARD_TOPICS">
+             ng-click="setActiveTab(TAB_NAME_TOPICS)"> Topics
```
Move "Topics" to its own line (and indent by 2 from the a tag). Ditto below.

aks681 replied:
Done.

seanlip commented on core/templates/dev/head/pages/topics_and_skills_dashboard/topics_and_skills_dashboard.html:
```diff
@@ -25,38 +25,35 @@ <h2 class="dashboard-title" translate="I18N_DASHBOARD_TOPICS_AND_SKILLS_DASHBOAR
       <blockquote class="intro-card">
-        <p class="intro-card-message" style="padding-top: 6px;"
-           translate="I18N_TOPICS_AND_SKILLS_DASHBOARD_INTRO_MESSAGE">
+        <p class="intro-card-message" style="padding-top: 6px;"> No topics or skills have been created yet.
```
Ditto, move "No topics ..." to its own line, indented from the p tag.

In general, if it's not of the form `<tag>...</tag>` on a single line, have a line break after the opening tag. /cc @apb7 @kevinlee12

aks681 replied:
Ok, done.

apb7 replied:
Added in Issue #5108.

seanlip commented on core/templates/dev/head/pages/topics_and_skills_dashboard/topics_list_directive.html:
```diff
@@ -1,97 +1,94 @@
 
 <style>
-  .topics-list .list-view-item {
+  topics-list .list-view-item {
```
Again do you need list-view-item?

aks681 replied:
Same here, when I removed .list-view-item, the styling was removed.

seanlip reviewed (approved):
Basically LGTM, please fix remaining comments and it's good to merge. (Most are small nits, but please make sure to catch all cases of the skill ID naming thing!)

aks681 committed:
made review changes
